File: web/src/engine/cli_test.ts
/**
 * Culinary Alchemy - CLI Test Runner
 * Validates combinations, practice exp, and skill unlocks in Node.js.
 */

import "../data/index";
import "../progression_config";
import { ProgressionEngine } from "./progression_engine";
import { CombinationEngine } from "./combination_engine";

function assert(condition: unknown, message: string): void {
  if (!condition) {
    console.error(`❌ FAIL: ${message}`);
    process.exit(1);
  }
  console.log(`✅ PASS: ${message}`);
}

function runTests(): void {
  console.log("=== STARTING CULINARY ENGINE TESTS ===");

  const progression = new ProgressionEngine(globalThis.PROGRESSION_CONFIG);
  const combination = new CombinationEngine(
    globalThis.DISCOVERABLE_ITEMS,
    globalThis.TRANSITION_INDEX
  );

  const itemsCount = Object.keys(globalThis.DISCOVERABLE_ITEMS).length;
  assert(itemsCount === 224, `Discoverable items count is ${itemsCount} (expected 224)`);

  const techniqueMap = globalThis.TRANSITION_INDEX.getTechniqueItemMap();
  assert(
    Array.isArray(techniqueMap.separate) && techniqueMap.separate.includes("berries"),
    "Transition index maps separate → affectable items including berries."
  );
  assert(
    globalThis.TRANSITION_INDEX.getTechniqueTransition("separate", "berries")?.input === "berries",
    "Indexed lookup resolves separate + berries without scanning all items."
  );
  assert(progression.getXP("smash") === 0, "Smash starts at 0 exp.");
  assert(progression.isUnlocked("smash") === true, "Smash is unlocked at start.");
  assert(progression.isUnlocked("pound") === false, "Pound starts locked.");
  assert(progression.getXP("combine") === 0, "Combine exp starts at 0.");
  assert(progression.isUnlocked("hand_mix") === false, "Hand Mix starts locked (needs combine exp).");
  assert(progression.getXP("separate") === 0, "Separate exp starts at 0.");
  assert(progression.isUnlocked("peel") === false, "Peel starts locked (needs separate exp).");
  assert(progression.isUnlocked("tear") === false, "Tear starts locked (needs separate exp).");

  const noSmashOnPrimal = combination.matchToolRecipe("tubers", "force", progression);
  assert(noSmashOnPrimal.success === false, "Force does not work on unsorted primal tubers.");

  const smashPotato = combination.matchToolRecipe("potato", "force", progression);
  assert(smashPotato.success === true, "Force turns potato into mashed potato.");
  assert(smashPotato.recipe?.result.id === "mashed_potato", "Force output is mashed potato.");

  const charApple = combination.matchToolRecipe("apple", "char", progression);
  assert(charApple.success === true, "Char turns apple into charred apple.");
  assert(charApple.recipe?.result.id === "charred_apple", "Char output is charred apple.");

  let xpNotif1 = progression.addXP("smash", 1);
  assert(progression.getXP("smash") === 1, "Smash exp is now 1.");
  assert(xpNotif1.leveledUp === false, "Pound is still locked (needs 3 exp).");

  let xpNotif2 = progression.addXP("smash", 2);
  assert(progression.getXP("smash") === 3, "Smash exp reached 3.");
  assert(xpNotif2.leveledUp === true, "Smash practicing triggers skill unlock.");
  assert(xpNotif2.newlyUnlockedSkills[0].id === "pound", "Pound is newly unlocked!");
  assert(progression.isUnlocked("pound") === true, "Pound is now fully unlocked.");

  const activeSmashTier = progression.getActiveTier("smash");
  assert(activeSmashTier?.id === "pound", "Active smash-path skill is now Pound.");

  const mergeResult = combination.matchCombinationRecipe(["seeds", "water"]);
  assert(mergeResult.success === true, "Seeds and water combine into sprouted seeds.");
  assert(mergeResult.recipe?.result.id === "sprouted_seeds", "Combine output is sprouted seeds.");

  const berryBrew = combination.matchCombinationRecipe(["strawberry", "spring_water"]);
  assert(berryBrew.success === true, "Strawberry and spring water combine into berry brew.");
  assert(berryBrew.recipe?.result.type === "recipe", "Berry brew is a finalized recipe.");

  progression.addXP("combine", 3);
  assert(progression.isUnlocked("hand_mix") === true, "Hand Mix unlocks at 3 combine exp.");

  progression.addXP("separate", 2);
  assert(progression.isUnlocked("peel") === true, "Peel unlocks at 2 separate exp.");

  const berryMatch = combination.matchToolRecipe("berries", "separate", progression, {
    discoveredIds: new Set()
  });
  assert(berryMatch.success === true, "Berries separate into the first undiscovered fruit.");
  assert(berryMatch.recipe?.results?.length === 1, "Berry separation yields one output per action.");
  assert(berryMatch.recipe?.result.id === "strawberry", "First separation yields strawberry.");

  const berryMatch2 = combination.matchToolRecipe("berries", "separate", progression, {
    discoveredIds: new Set(["strawberry"])
  });
  assert(berryMatch2.recipe?.result.id === "raspberry", "Second separation yields raspberry.");

  const berryExhausted = combination.matchToolRecipe("berries", "separate", progression, {
    discoveredIds: new Set(["strawberry", "raspberry", "blueberry", "blackberry", "smashed_berries"])
  });
  assert(berryExhausted.success === false, "Separation fails when all berry types are discovered.");

  const fruitMatch = combination.matchToolRecipe("fruits", "separate", progression, {
    discoveredIds: new Set()
  });
  assert(fruitMatch.success === true, "Fruits separate into the first undiscovered fruit.");
  assert(fruitMatch.recipe?.results?.length === 1, "Fruit separation yields one output per action.");
  assert(fruitMatch.recipe?.result.id === "apple", "First fruit separation yields apple.");

  const fruitMatch2 = combination.matchToolRecipe("fruits", "separate", progression, {
    discoveredIds: new Set(["apple"])
  });
  assert(fruitMatch2.recipe?.result.id === "banana", "Second fruit separation yields banana.");

  const fruitExhausted = combination.matchToolRecipe("fruits", "separate", progression, {
    discoveredIds: new Set([
      "apple", "banana", "orange", "grape", "pear",
      "watermelon", "mango", "pineapple", "lemon", "peach"
    ])
  });
  assert(fruitExhausted.success === false, "Separation fails when all fruit types are discovered.");

  const waterMatch = combination.matchToolRecipe("water", "separate", progression, {
    discoveredIds: new Set()
  });
  assert(waterMatch.success === true, "Water separates into the first undiscovered type.");
  assert(waterMatch.recipe?.result.id === "spring_water", "First water separation yields spring water.");

  const rootsMatch = combination.matchToolRecipe("roots", "separate", progression, {
    discoveredIds: new Set()
  });
  assert(rootsMatch.success === true, "Roots separate one at a time.");
  assert(rootsMatch.recipe?.result.id === "carrot", "First root separation yields carrot.");

  const milestones = progression.checkMilestoneUnlocks(4);
  assert(milestones.length === 1, "First milestone unlocks at 3 discoveries.");
  assert(milestones[0].unlocks.includes("seeds"), "First milestone unlocks seeds.");

  console.log("=== ALL CULINARY ENGINE TESTS PASSED SUCCESSFULLY ===");
}

runTests();

File: archive/core/tests/cli_test.rs
use culinary_core::{CombinationEngine, GameBundle, ProgressionEngine};
use std::collections::HashSet;
use std::path::PathBuf;

fn assets_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("assets")
}

#[test]
fn engine_matches_typescript_cli_tests() {
    let bundle = GameBundle::load(&assets_dir()).expect("load assets");
    assert_eq!(bundle.discoverable.len(), 66);

    let mut progression = ProgressionEngine::new(bundle.progression.clone(), None);
    let engine = CombinationEngine::new(&bundle.discoverable, &bundle.index);

    assert!(bundle.index.by_technique.contains_key("separate"));
    assert!(bundle
        .index
        .by_technique
        .get("separate")
        .unwrap()
        .contains_key("berries"));

    assert_eq!(progression.get_xp("smash"), 0);
    assert!(progression.is_unlocked("smash"));
    assert!(!progression.is_unlocked("pound"));
    assert!(!progression.is_unlocked("hand_mix"));
    assert!(!progression.is_unlocked("peel"));

    let no_smash = engine.match_technique("tubers", "smash", &mut progression, &HashSet::new());
    assert!(!no_smash.success);

    let smash_potato = engine.match_technique("potato", "smash", &mut progression, &HashSet::new());
    assert!(smash_potato.success);
    assert_eq!(smash_potato.result_id.as_deref(), Some("mashed_potato"));

    let char_apple = engine.match_technique("apple", "char", &mut progression, &HashSet::new());
    assert!(char_apple.success);
    assert_eq!(char_apple.result_id.as_deref(), Some("charred_apple"));

    let xp1 = progression.add_xp("smash", 1);
    assert_eq!(progression.get_xp("smash"), 1);
    assert!(xp1.is_empty());

    let xp2 = progression.add_xp("smash", 2);
    assert_eq!(progression.get_xp("smash"), 3);
    assert_eq!(xp2, vec!["pound"]);
    assert!(progression.is_unlocked("pound"));

    let merge = engine.match_combine(&["seeds".into(), "water".into()]);
    assert!(merge.success);
    assert_eq!(merge.result_id.as_deref(), Some("sprouted_seeds"));

    let berry_brew = engine.match_combine(&["strawberry".into(), "spring_water".into()]);
    assert!(berry_brew.success);
    assert!(bundle.discoverable["berry_brew"].is_recipe());

    progression.add_xp("combine", 3);
    assert!(progression.is_unlocked("hand_mix"));

    progression.add_xp("separate", 2);
    assert!(progression.is_unlocked("peel"));

    let berry1 = engine.match_technique(
        "berries",
        "separate",
        &mut progression,
        &HashSet::new(),
    );
    assert!(berry1.success);
    assert_eq!(berry1.result_ids.len(), 1);
    assert_eq!(berry1.result_id.as_deref(), Some("strawberry"));

    let mut discovered = HashSet::from(["strawberry".to_string()]);
    let berry2 = engine.match_technique("berries", "separate", &mut progression, &discovered);
    assert_eq!(berry2.result_id.as_deref(), Some("raspberry"));

    discovered.extend(["raspberry", "blueberry", "blackberry"].map(String::from));
    let berry_exhausted =
        engine.match_technique("berries", "separate", &mut progression, &discovered);
    assert!(!berry_exhausted.success);

    let fruit1 = engine.match_technique("fruits", "separate", &mut progression, &HashSet::new());
    assert!(fruit1.success);
    assert_eq!(fruit1.result_id.as_deref(), Some("apple"));

    let mut fruits_discovered = HashSet::from(["apple".to_string()]);
    let fruit2 = engine.match_technique("fruits", "separate", &mut progression, &fruits_discovered);
    assert_eq!(fruit2.result_id.as_deref(), Some("banana"));

    fruits_discovered.extend(
        [
            "banana", "orange", "grape", "pear", "watermelon", "mango", "pineapple", "lemon",
            "peach",
        ]
        .map(String::from),
    );
    let fruit_exhausted =
        engine.match_technique("fruits", "separate", &mut progression, &fruits_discovered);
    assert!(!fruit_exhausted.success);

    let water = engine.match_technique("water", "separate", &mut progression, &HashSet::new());
    assert!(water.success);
    assert_eq!(water.result_id.as_deref(), Some("spring_water"));

    let roots = engine.match_technique("roots", "separate", &mut progression, &HashSet::new());
    assert!(roots.success);
    assert_eq!(roots.result_id.as_deref(), Some("carrot"));
}

#[test]
fn runtime_unlocks_achievements_from_shared_rules() {
    use culinary_core::GameRuntime;

    let bundle = GameBundle::load(&assets_dir()).expect("load assets");
    assert_eq!(bundle.achievements.len(), 18);
    assert!(bundle.achievement_rules.contains_key("first_combine"));

    let mut rt = GameRuntime::new(bundle);
    let result = rt.apply_combine("seeds", "water");
    assert!(result.success);
    assert!(
        result.new_achievement_ids.contains(&"first_combine".to_string()),
        "combine_success flag should unlock first_combine"
    );
    assert!(rt.achievement_unlocks().contains_key("first_combine"));
}
